### PR TITLE
u3d/dependencies: add command to install Linux dependencies

### PR DESCRIPTION
**Pull request by niezbop**
Before installing Unity on Linux, the installer will now install its required dependencies.

### CODE REVIEW

**Comment by lacostej** [comment]
A few things make me uncertain about how the current proposal

1. we force install of the dependencies everytime we try to install. Dependencies are (to my knowledge) for **running** the editor. Do we need them for installation?
2. we force install of the dependencies everytime we try to install and regardless if the dependencies are there already
3. we fail hard if we don't recognize the platform (deb/yum-rpm). This prevents installing unity completely on those systems
4. the check dev vs rpm doesn't allow us to ensure you are on the right platform (if for example you can install dpkg on a rpm based platform). 
5. Fedora migrated to dnf. Not sure how that works there.

I would like this to be tested heavily on various systems. Maybe implemented as a separate command (`u3d install-editor-deps [--type deb/rpm]` ?

**Comment by niezbop** [comment]
I agree it is a peculiar feature to have. Missing libraries did indeed have no effect on whether the installer completed its job.

1. Short answer no, but what good is it to install something that won't work?

2. That's why I thought that validation for each library would be better, as a user could specify to skip some of them, knowing they're already present, or not willing to update them.
Alternatively, we could either keep track of what has been installed or try to fetch installed libraries before trying to install them.

3. return and UI.error then?

4. I'm not sure there is a really better way, it seems to be a well known check. The usecases of people having installed dpkg on a rpm-based os (which has no point as .deb files are Debian specific) is too small (in my opinion at least) to have an impact on this. Then again, there may be a far better check that I'm not aware of! Nothing is written in stone anyways, so my current check could be swapped easily.

5. I'll try to find a solution for that

This could be a good idea indeed. Especially if we consider the added dependencies needed to build to WebGL or to Tizen/Android, which could be specified in such a command as well.

**Comment by lacostej** [comment]
Notes for self: detect if something is to be installed on Debian derivative:
```
$ sudo apt-get install -s maven | grep "^Inst"  | wc -l
84
$ sudo apt-get install -s libxi6 | grep "^Inst"  | wc -l
0
```
Is there a better way?

**Comment by lacostej** [comment]
> Short answer no, but what good is it to install something that won't work?

I am not sure I want the question: "do you want to install?" everytime.
An alternative would be to check if there are already some linux versions installed. If not, we could ask. But I don't like to mix the other version presents detection code down the installer. I would prefer an option to the method being passed down (`auto_install_deps: true/false`). WDYT?

> That's why I thought that validation for each library would be better, as a user could specify to
> skip some of them, knowing they're already present, or not willing to update them.
> Alternatively, we could either keep track of what has been installed or try to fetch installed 
> libraries before trying to install them.

I prefer the current alternative where I don't want to install all of them. The package manager knows if the packages are there or not. We are not checking version number either. So it's a one shot attempt at getting everything in place.

> return and UI.error then?

If it's explicit installation by the user, fail. If it's automatic try installing them, error and return.

> I'm not sure there is a really better way, it seems to be a well known check. The usecases of
> people having installed dpkg on a rpm-based os (which has no point as .deb files are Debian
> specific) is too small (in my opinion at least) to have an impact on this. Then again, there may be
> a far better check that I'm not aware of! Nothing is written in stone anyways, so my current check
> could be swapped easily.

I agree it's a small case. I would prefer maybe something like

if not packaging system specified
  try auto-detecting
  if can't
     ....


> I'll try to find a solution for that

OK

I still think we could have a separate action for installing the deps from the CLI to cover the cases.

**Comment by lacostej** [comment]
I propose we merge this in the following way.

* we add the missing libpq5
* we add an API so that we can call programmatically something like
  `U3d::LinuxDependencies.install(type: :rpm)`
* We don't integrate the call at install time
* we optionally provide a command. But I can live without this right now.

My objective is to be able to test this on master easily. WDYT?

**Comment by niezbop** [comment]
Seems really good to me! I'll work on doing that.

File: lib/u3d/commands.rb [hunk 1]
```diff
@@ -129,6 +129,15 @@ def install(args: [], options: {})
         Installer.install_modules(files, definition.version, installation_path: options[:installation_path])
       end
 
+      def install_linux_dependencies(options: {})
```
Comment (lacostej) [inline lib/u3d/commands.rb]: `install_dependencies` ?

File: lib/u3d/commands_generator.rb [hunk 2]
```diff
@@ -148,6 +148,15 @@ def run
         end
       end
 
+      command :dependencies do |c|
+        c.syntax = 'u3d dependencies [-p | --package_manager <package manager>]'
+        c.description = 'Installs Unity dependencies. [Linux only]'
```
Comment (lacostej) [inline lib/u3d/commands_generator.rb]: summary + description. We can be more verbose on the description part. In particular
*  detail how it works for detecting the package_manager, etc.
* link to Linux forum post
* explain how we deal with things like `debconf (>= 0.5) | debconf-2.0`
* the fact we only implement the default deps, and not the ones for platforms (WebGL, etc)

File: lib/u3d/commands_generator.rb [hunk 3]
```diff
@@ -148,6 +148,15 @@ def run
...
+        c.option '-p', '--package_manager STRING', String, 'Specify which package manager to use (aptitude, rpm...)'
```
Comment (lacostej) [inline lib/u3d/commands_generator.rb]: option is unused right now, correct? If so remove it.

Comment (niezbop) [reply lib/u3d/commands_generator.rb]: Proposed implementation, currently in this PR:

```
def self.install(prefix: nil)
      unless prefix
        if `which dpkg` != ''
          prefix = 'apt-get -y install'
        elsif `which rpm` != ''
          prefix = 'yum -y install'
        else
          raise 'Cannot install dependencies on your Linux distribution'
        end
      end
      if UI.interactive?
        return unless UI.confirm "Install dependencies? (#{DEPENDENCIES.length} dependency(ies) to install)"
      end
      U3dCore::CommandExecutor.execute(command: "#{prefix} #{DEPENDENCIES.join(' ')}", admin: true)
    end
```

File: lib/u3d/commands_generator.rb (hunk 2 above)
Comment (niezbop) [reply lib/u3d/commands_generator.rb]: Ok

File: lib/u3d/commands_generator.rb [hunk 4]
```diff
@@ -148,6 +148,26 @@ def run
         end
       end
 
+      command :dependencies do |c|
+        c.syntax = 'u3d dependencies [-p | --package_manager <package manager>]'
+        c.summary = 'Installs Unity dependencies. [Linux only]'
+        c.description = %(
+          #{c.summary}
```
Comment (lacostej) [inline lib/u3d/commands_generator.rb]: Do we need to remove the front spaces here? Not sure how `%(` behaves.

File: lib/u3d/commands_generator.rb [hunk 5]
```diff
@@ -148,6 +148,26 @@ def run
...
+
+          Regarding the package manager: if none if specified, u3d will try to assume it by seeing which one.s is.are installed.
```
Comment (lacostej) [inline lib/u3d/commands_generator.rb]: various typos.

File: lib/u3d/commands_generator.rb [hunk 6]
```diff
@@ -148,6 +148,26 @@ def run
...
+          Only apt-get and yum are tested right now.
+
+          If several package manager are installed on your machine simultaneously, we recommend that you specify it.
+
+          Regarding the dependencies themselves: only dependencies for the editor are installed. WebGL, Android and Tizen require others that you will have to install manually.
+          More on that: https://forum.unity3d.com/threads/unity-on-linux-release-notes-and-known-issues.350256/
+                  )
+        c.option '-p', '--package_manager STRING', String, 'Specify which package manager to use (apt-get, yum...)'
```
Comment (lacostej) [inline lib/u3d/commands_generator.rb]: do you really want to use it with `-p yum` ?

Comment (niezbop) [reply lib/u3d/commands_generator.rb]: Not a great option, but it was the least bad I could think of. Any suggestion?

File: lib/u3d/commands_generator.rb [hunk 7]
```diff
@@ -148,6 +148,26 @@ def run
...
+      command :dependencies do |c|
+        c.syntax = 'u3d dependencies [-p | --package_manager <package manager>]'
+        c.summary = 'Installs Unity dependencies. [Linux only]'
+        c.description = %(
+#{c.summary}
+
+Regarding the package manager: if none is specified, u3d will try to assume it by testing which one.s is.are installed.
+Only apt-get and yum are tested right now.
+
+If several package manager are installed on your machine simultaneously, we recommend that you specify the one you want to use.
```
Comment (lacostej) [inline lib/u3d/commands_generator.rb]: typo: managers

File: lib/u3d/commands_generator.rb [hunk 8]
```diff
@@ -148,6 +148,26 @@ def run
...
+
+Regarding the dependencies themselves: only dependencies for the editor are installed. WebGL, Android and Tizen require others that you will have to install manually.
```
Comment (lacostej) [inline lib/u3d/commands_generator.rb]: will that work if user doesn't specify -p "apt-get -y" ?

I.e. do we manage input properly?

Comment (niezbop) [reply lib/u3d/commands_generator.rb]: Not working properly right now, I'm correcting that

File: lib/u3d/commands_generator.rb [hunk 9]
```diff
@@ -148,6 +148,23 @@ def run
...
+        c.summary = 'Installs Unity dependencies. [Linux only]'
+        c.description = %(
+#{c.summary}
+
+Regarding the package manager: if dpkg is installed, u3d uses apt-get else if rpm is installed yum is used. If none of them is insalled, fails.
+
+Regarding the dependencies themselves: only dependencies for the editor are installed. WebGL, Android and Tizen require others that you will have to install manually.
+More on that: https://forum.unity3d.com/threads/unity-on-linux-release-notes-and-known-issues.350256/
+                  )
+        c.option '-p', '--package_manager STRING', String, 'Specify which package manager to use (apt-get, yum...)'
```
Comment (lacostej) [inline lib/u3d/commands_generator.rb]: remove

File: lib/u3d/commands_generator.rb [hunk 10]
```diff
@@ -148,6 +148,23 @@ def run
         end
       end
 
+      command :dependencies do |c|
+        c.syntax = 'u3d dependencies [-p | --package_manager <package manager>]'
```
Comment (lacostej) [inline lib/u3d/commands_generator.rb]: update

**Pull request merged**